#summary Tutorial for MyDAS - Intermediate level
#labels Phase-Support
= MyDAS 1.6 Tutorial =

== Second Part: Developing your own Data Source ==
Now that you have been a real DAS source working on MyDAS lets get closer to a more interesting example: what if you have your data in an in-house format? and therefore the templates/examples don't apply for your data source?

=== Example File ===
The file of this example was built using information from the ensembl database homo_sapiens_core_56_37a. The file gives positional information of genes, transcripts and exons in the chromosome 5 organized in a separated by pipes `|` like this:
{{{
|chr|gene_id|gene_start|gene_end|trascript_id|transcript_start|transcript_end|exon_id|exon_start|exon_end|
}}}
You can see a sample of this file below.
{{{
|5|ENSG00000153404|140373|190087|ENST00000398036|140373|157131|ENSE00001648483|156888|157131|
|5|ENSG00000153404|140373|190087|ENST00000398036|140373|157131|ENSE00001135995|156186|156325|
|5|ENSG00000153404|140373|190087|ENST00000398036|140373|157131|ENSE00001136002|155460|155558|
|5|ENSG00000153404|140373|190087|ENST00000398036|140373|157131|ENSE00001136007|154990|155106|
|5|ENSG00000153404|140373|190087|ENST00000398036|140373|157131|ENSE00001136016|151628|151714|
|5|ENSG00000153404|140373|190087|ENST00000398036|140373|157131|ENSE00001136025|144942|145035|
|5|ENSG00000153404|140373|190087|ENST00000398036|140373|157131|ENSE00001136029|143495|143618|
}}}


=== Analyzing the Data ===
Firstly, we know from this file the information that our data source has to provide and from them that there will be just four types of features in this data source:
  * chromosome: In our example will be always the same(5) however we can create our source to support files with information from more than one chromosome. This type should look in DAS like: 
{{{ 
<TYPE id="Chromosome" cvId="SO:0000340">Chromosome</TYPE> 
}}}
  * genes: The file give us information about the start and end of the genes plus its ID. From there we have another type in our data source:
{{{
<TYPE id="Gene" cvId="SO:0000704">Gene</TYPE>
}}}
  But also we have the id field of each feature of this type and the tags start and end can be also deduced from there.
  * transcripts: As in the case of genes the information provided is the id, start and stop. So we will need a new DAS type:
{{{
<TYPE id="Transcript" cvId="SO:0000673">Transcript</TYPE>
}}}
  * exons: Same analysis as above:
{{{
<TYPE id="Exon" cvId="SO:0000147">Exon</TYPE>
}}}
Note that the numbers at cvId are valid Ontology Ids from the Sequence Ontology.

The other information that we have implicit in this file is that there is a hierarchy in those components. A chromosome contains genes, a gene contains transcripts, an transcript contains exons. 

For this demo we can assume some values for all the annotations, the method will be
{{{
<METHOD id="not_recorded" cvId="ECO:0000037">not_recorded</METHOD> 
}}}
And given that we don't know the score, phase and orientation, we will assume those values as not applicable.